### PR TITLE
[REVIEW] Don't shallow copy index in as_index() unless necessary

### PR DESCRIPTION
**Pull request by shwina**
An optimization that prevents us from having to do index comparisons unnecessarily.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/core/index.py [hunk 1]
```diff
@@ -2701,14 +2701,14 @@ def as_index(arbitrary, **kwargs):
...
     kwargs = _setdefault_name(arbitrary, **kwargs)
     if isinstance(arbitrary, cudf.MultiIndex):
         return arbitrary
     elif isinstance(arbitrary, Index):
+        if arbitrary.name == kwargs["name"]:
+            return arbitrary
         idx = arbitrary.copy(deep=False)
-        idx.rename(kwargs.get("name"), inplace=True)
-        return idx
+        return idx.rename(kwargs["name"], inplace=True)
```
Comment (kkraus14) [inline python/cudf/cudf/core/index.py:2711]: Doesn't `idx.rename(..., inplace=True)` return `None`?

Comment (shwina) [reply python/cudf/cudf/core/index.py:2711]: Oh -- right

Comment (shwina) [reply python/cudf/cudf/core/index.py]: fixed

**Review by kkraus14 — APPROVED** [review]

File: python/cudf/cudf/core/join/join.py [hunk 2]
```diff
@@ -512,6 +512,10 @@ def typecast_libcudf_to_output(self, output, output_dtypes):
                     output._index._data[
                         index_col_lbl
                     ] = self._build_output_col(index_col, index_dtype)
+            # reconstruct the Index object as the underlying data types
+            # have changed:
+            output._index = cudf.core.index.Index._from_table(output._index)
```
Comment (shwina) [inline python/cudf/cudf/core/join/join.py:517]: Looks like we were implicitly relying on this happening from before

**Review by kkraus14 — APPROVED** [review]

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
Looks like there's a couple of dask-cudf test failures: https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fcudf%2Fprb%2Fcudf-gpu-build/detail/cudf-gpu-build/32489/tests

**Review by kkraus14 — APPROVED** [review]

**Pull request merged**